sqlx-core/src/net/socket/mod.rs

```diff
@@ -182,10 +183,87 @@ impl<S: Socket + ?Sized> Socket for Box<S> {
     }
 }
 
+#[derive(Debug, Clone, Copy)]
+pub struct TcpKeepalive {
+    pub time: Option<Duration>,
+    pub interval: Option<Duration>,
+    pub retries: Option<u32>,
+}
+
+impl TcpKeepalive {
+    /// Returns a new, empty set of TCP keepalive parameters.
+    pub const fn new() -> TcpKeepalive {
+        TcpKeepalive {
+            time: None,
+            interval: None,
+            retries: None,
+        }
+    }
+
+    /// Set the amount of time after which TCP keepalive probes will be sent on
+    /// idle connections.
+    ///
+    /// This will set `TCP_KEEPALIVE` on macOS and iOS, and
+    /// `TCP_KEEPIDLE` on all other Unix operating systems, except
+    /// OpenBSD and Haiku which don't support any way to set this
+    /// option. On Windows, this sets the value of the `tcp_keepalive`
+    /// struct's `keepalivetime` field.
+    ///
+    /// Some platforms specify this value in seconds, so sub-second
+    /// specifications may be omitted.
+    pub const fn with_time(self, time: Duration) -> Self {
+        Self {
+            time: Some(time),
+            ..self
+        }
+    }
+
+    /// Set the value of the `TCP_KEEPINTVL` option. On Windows, this sets the
+    /// value of the `tcp_keepalive` struct's `keepaliveinterval` field.
+    ///
+    /// Sets the time interval between TCP keepalive probes.
+    ///
+    /// Some platforms specify this value in seconds, so sub-second
+    /// specifications may be omitted.
+    pub const fn with_interval(self, interval: Duration) -> Self {
+        Self {
+            interval: Some(interval),
+            ..self
+        }
+    }
+
+    /// Set the value of the `TCP_KEEPCNT` option.
+    ///
+    /// Set the maximum number of TCP keepalive probes that will be sent before
+    /// dropping a connection, if TCP keepalive is enabled on this socket.
+    pub const fn with_retries(self, retries: u32) -> Self {
+        Self {
+            retries: Some(retries),
+            ..self
+        }
+    }
+
+    /// Convert `TcpKeepalive` to `socket2::TcpKeepalive`.
+    pub const fn socket2(self) -> socket2::TcpKeepalive {
+        let mut ka = socket2::TcpKeepalive::new();
+        if let Some(time) = self.time {
+            ka = ka.with_time(time);
+        }
+        if let Some(interval) = self.interval {
+            ka = ka.with_interval(interval);
```

Comment:
This setter isn't defined on certain platforms (shown here is the field which is `cfg`d out for platforms where it's _not_ supported): https://docs.rs/socket2/latest/x86_64-apple-darwin/src/socket2/lib.rs.html#461

Reply:
I am trying to make changes! But after I have added all those conditional compilation `cfg`s, it's seems like that a bespoke `TcpKeepalive` is not necessary since the definition is (nearly) the same as that in `socket2`. We can use `socket2::TcpKeepalive` directly.
I chose to write a self-defined `TcpKeepalive` because @CommanderStorm suggested that `TcpKeepalive` could be `Copy` but now it seems `Copy` is not necessary. Could we just use `socket2::TcpKeepalive` and expose `tcp_keepalive_time()`  on the `ConnectOptions` builders?
( I'm not going to reply under every change-request since it's annoying to you :D )
Thank you for your reviewing!

Reply:
I do **not** want to export `socket2:TcpKeepalive` because that then becomes part of our public API. We would not be able to upgrade `socket2` without it being a breaking change.

Reply:
Okay, I have made changes according to your change-requests.